symengine/polynomial.cpp

```diff
@@ -296,9 +314,253 @@ RCP<const UnivariateIntPolynomial> mul_poly(RCP<const UnivariateIntPolynomial> a
     }
 
     if (sign == -1)
-        return neg_poly(*UnivariateIntPolynomial::from_vec(a->var_, v));
+        return neg_poly(*UnivariateIntPolynomial::from_vec(var, v));
     else
-        return UnivariateIntPolynomial::from_vec(a->var_, v);
+        return UnivariateIntPolynomial::from_vec(var, v);
+}
+
+
+UnivariatePolynomial::UnivariatePolynomial(const RCP<const Symbol> &var, const unsigned int &degree, const map_int_Expr&& dict) :
+     degree_{degree}, var_{var}, dict_{std::move(dict)} {
+    SYMENGINE_ASSERT(is_canonical(degree_, dict_))
+}
+
+UnivariatePolynomial::UnivariatePolynomial(const RCP<const Symbol> &var, const std::vector<Expression> &v) : var_{var} {
+    for (unsigned int i = 0; i < v.size(); i++)
+        if (v[i] != 0)
+            dict_add_term(dict_, v[i], i);
+    degree_ = v.size() - 1;
+    SYMENGINE_ASSERT(is_canonical(degree_, dict_))
+}
+
+bool UnivariatePolynomial::is_canonical(const unsigned int &degree_, const map_int_Expr& dict) const {
+    if (var_->get_name() == "")
+        if (!(dict.empty() or dict.size() == 1))
```

Comment:
`if (!(dict.empty() or (dict.size() == 1 and dict.begin()->first == 0)))` doesn't give me any warnings